cli/src/commands/rebase.rs

```diff
@@ -665,32 +665,33 @@ fn move_commits(
             .commits(mut_repo.store())
             .try_collect()?;
 
-    // Commits in the target set should only have other commits in the set as
-    // parents, except the roots of the set, which persist their original
-    // parents.
+    // Compute the parents of all commits in the connected target set,
+    // allowing only commits in the target set as parents.
     // If a commit in the set has a parent which is not in the set, but has
     // an ancestor which is in the set, then the commit will have that ancestor
```

Comment:
A better description is "If a commit in the target set has a parent which is not in the target set but has an ancestor in the target set, the parent's closest ancestors which are in the target set will be added to the commit's parents."

In the case of your example, `f` would have parents `d` (direct parent in the target set), as well as `b` and `c` (`e`'s parents which are in the target set).

Reply:
"Closest ancestors" is not well-defined either. Let me add a commit `z` (not a target commit) before `f`:

```
@    f
○    z
├─╮
│ ○    e
│ ├─╮
│ │ ○  c
○ │ │  d
├─╯ │
○   │  b
├───╯
○  a
◆
```

So a commit in the target set `f` has a parent that is not in the target set `z`. What are the parent's (that is, `z`'s) closest ancestors? It could arguably be (1) `d b c` (as you describe) or (2) `d c` (`b` is excluded because `d` is indisputably closer than `b`).

After some thought, the best way I could describe both would be:
(1) any ancestor of `f` that is a target commit and whose ancestry path (from `f`) does not go through another target commit
(2) the heads of the (the intersection of the ancestors of `f` and the target set)

Of the two, I prefer (2) because it reuses terminology from "Revset language", which means that it's easier to understand and test. (Putting myself in the user's shoes, this is a really contrived example, so I think either is reasonable, as long as we are consistent.) But I think this means rewriting the code (as far as I can tell, the code both before and after this PR implements (1)). I can't think of an efficient algorithm to do (2) off the top of my head, unfortunately. (Git does something similar during fetch negotiation, but the output is one big set of "bottom" commits; here, we need one set of "bottom" commits per target commit, so we cannot directly use Git's algorithm.)